### PR TITLE
table: drop trailing white spaces in the last cell

### LINKED ISSUE #18415: suppress end of line whitespace in table output
C.f https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=980820

Please clean up the output of all systemd commands by removing extra
whitespaces at EOL ::

    $ timedatectl status | cat -vTE

                   Local time: Sat 2021-01-30 17:46:26 EET $
               Universal time: Sat 2021-01-30 15:46:26 UTC $
                     RTC time: Sat 2021-01-30 15:46:26     $
                    Time zone: Europe/Helsinki (EET, +0200)$
    System clock synchronized: yes                         $
                  NTP service: active                      $
              RTC in local TZ: no                          $

Debian policy shows ::

    systemd:
      Installed: 247.2-5

NOTE: The output was clean in previous versions ::

    $ timedatectl status | cat -vTE
          Local time: Sat 2021-01-30 17:46:21 EET$
      Universal time: Sat 2021-01-30 15:46:21 UTC$
            RTC time: Sat 2021-01-30 15:46:21$
           Time zone: Europe/Helsinki (EET, +0200)$
     Network time on: yes$
    NTP synchronized: yes$
     RTC in local TZ: no$

Debian policy shows ::

    systemd:
      Installed: 232-8

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #18415.

### CODE REVIEW

File: src/shared/format-table.c [hunk 1]
```diff
@@ -2155,6 +2155,9 @@ int table_print(Table *t, FILE *f) {
                                                 if (!aligned)
                                                         return -ENOMEM;
 
+                                                if (j == display_columns - 1 && d->align_percent == 0)
+                                                        delete_trailing_chars(aligned, NULL);
```
Comment (poettering) [inline src/shared/format-table.c:2265]: hmm, sounds pointless to first add the space, and then drop it again.  The space should not be added at all in the first place. 

moreover i guess the whitespace should also be suppressed for any alignment

**Review by poettering — CHANGES_REQUESTED** [review]
hmm, this is problematic in case a cell uses an ansi color with background. i am pretty sure in that case the full cell should be highlighted with the bg color, not just the text.

that's the reason I didn't bother stripping the whitespace in the first place. i.e. figuring out whether some ansi sequence induces a background color or not was something i didn't want to get into

**Comment by poettering** [comment]
> that's the reason I didn't bother stripping the whitespace in the first place. i.e. figuring out whether some ansi sequence induces a background color or not was something i didn't want to get into

maybe an easy way out is to leave the spaces in when color is enabled, and only suppress it when color is turned off. This would then make thing things look good still on a terminal, but when piped to some pipeline we'll suppres the trailing space, since color is automatically turned off there

**Comment by yuwata** [comment]
@poettring Thank you for the suggestion. Updated. PTAL.

File: src/shared/format-table.c [hunk 2]
```diff
@@ -2258,6 +2258,11 @@ int table_print(Table *t, FILE *f) {
                                                 if (!aligned)
                                                         return -ENOMEM;
 
+                                                /* Drop trailing white spaces when color is not set. */
+                                                if (j == display_columns - 1 &&
+                                                    (!colors_enabled() || !table_data_color(d)))
```
Comment (poettering) [inline src/shared/format-table.c]: hmm, i think this also needs:

```c
… &&
(!urlify_enabled() || !d->url))
```

**Comment by yuwata** [comment]
@poettering Updated. URLification is also checked. PTAL.

**Pull request merged**